### DISCUSSION TITLE
Is Spack able to include a package with mainly script files and a particular workflow for running?

### DISCUSSION THREAD

**Discussion by disheng222** [spack/spack#20521] [Q&A] [answered]
Hi, 
Let me introduce a specific toolkit we developed first and then put forward my question. 

We have put our library Z-checker in Spack: Z-checker is a data compression assessment library including many useful analysis implementations. As for users, however, in order to the compression assessment, they often also need to install compressors or they can also use their own compressors in the z-checker assessment. 

Simply installing those dependent compressor packages doesn't work because linking them together also requires some specific scripts and operations. So, we developed **Z-checker-installer** to compile some individual code files with 'main function' using particular custom Makefiles and do a set of configurations (e.g., copy some scripts to some directories) particularly, so as to link the compressors with Z-checker. That is, Z-checker-instlaller is rather like an execution environment controller to aggregate multiple libraries/analysis/scripts together. 

For the installation, the user just needs to run _z-checker-installer.sh_ to install everything for the assessment environment. 
For the usage, the user just needs to run createZCCase.sh to create a directory (called workspace) which will collect many scripts, and then use runZCCase.sh to do a series of compression assessments. All these are controlled by bash scripts.

My question is: 
Can we put z-checker-installer in Spack? 
(1) If yes, how to write the package.py? 
(2) How to run the Z-checker installer if it's installed by Spack.

Thanks.

Reference: 
Z-checker: https://github.com/CODARcode/z-checker
Z-checker-installer: https://github.com/CODARcode/z-checker-installer

Sheng

**Comment by becker33** [comment, marked as answer]
Yes, you can put z-checker into Spack.

The package.py file in Spack can be specialized to particular build systems, but the basic `Package` type is always an option. If your package inherits from Package, you can simply define an `install` method with the following signature.

```
def install(self, spec, prefix):
```

that method will install the package into the prefix, parameterized by `spec`.

For example, if there were one variant `bar` on package `foo`, you may have an install method that looks something like

```
def install(self, spec, prefix):
    installer = which('./foo-installer.sh')
    args = ['--prefix=%s' % prefix]
    if spec.satisfies('+bar'):
        args.append('--enable-foo')
    installer(*args)
```
